Dev

apmypb opened the pull request:
We can now specify the sizes of chunks to write to the file. 
I set the default to contiguous data.
Meaning, if someone wants the data to be chunked, he has to write 
data with an additional chunk_size argument:
```
data = rand(100)
chunk_size = 100
lhd["data", chunk_size] = data
```
where `lhd` is a LHDataStore

fhagemann commented:
Is it possible to read old files in with this or might this result in errors?

fhagemann commented:
I tested your changes using this script:
```julia
using LegendHDF5IO
using Test

using ArraysOfArrays
using DataFrames
using RadiationDetectorSignals
using StatsBase
using TypedTables
using Unitful


filename = "LegendHDF5IO_test.lh5"

A = randn(1000)
B = fill(true, 100)
F = Float64(6.5)
W = RDWaveform(range(0,100,length=100), rand(100))
U = fill(1.0u"mm", 100)
V = 1.0u"pF"
AW = ArrayOfRDWaveforms([W])
AS = ArrayOfSimilarArrays([deepcopy(U) for _ in 1:5])
VV = VectorOfVectors([fill(1.0u"mV", i) for i in 1:5])
hist = fit(Histogram, A)
nt = (A = A, B = B, F = F)
S = "TestString"
SYM = :Symbol

# Create the LH5 file
LHDataStore(filename, "w") do h
    h["Array"] = A
    h["BoolArray"] = B
    h["Histogram"] = hist
    h["Float"] = F
    h["ArrayOfWaveforms"] = AW
    h["ArrayWithUnits"] = U
    h["ArrayOfSimilarArrays"] = AS
    h["VectorOfVectors"] = VV
    h["Value"] = V
    h["NamedTuple"] = nt
    h["String"] = S
    # h["Waveform"] = W
    # h["Symbol"] = SYM
end

# read the data back in
@testset "LHDataStore read-in" begin
    LHDataStore(filename, "r") do h_in
        @test h_in["Array"] == A
        @test h_in["Array"][:] isa typeof(A)
        @test h_in["BoolArray"] == B
        @test_broken h_in["BoolArray"][:] isa typeof(B)
        @test h_in["Histogram"] == hist
        @test h_in["Histogram"] isa typeof(hist)
        @test h_in["Float"] == F
        @test h_in["Float"] isa typeof(F)
        @test_broken h_in["ArrayOfWaveforms"] == AW # requires a Colon somehow
        @test h_in["ArrayOfWaveforms"][:] == AW
        @test h_in["ArrayOfWaveforms"][:] isa typeof(AW)
        @test h_in["ArrayWithUnits"] == U
        @test h_in["ArrayWithUnits"][:] isa typeof(U)
        @test h_in["ArrayOfSimilarArrays"] == AS
        @test h_in["ArrayOfSimilarArrays"][:] isa typeof(AS)
        @test h_in["VectorOfVectors"] == VV
        @test h_in["VectorOfVectors"][:] isa typeof(VV)
        @test h_in["Value"] == V
        @test h_in["Value"] isa typeof(V)
        @test h_in["NamedTuple"] == nt
        @test_broken h_in["NamedTuple"] isa typeof(nt) # What's the command to read the Arrays into memory?
        @test h_in["String"] == S
        @test h_in["String"] isa typeof(S)
        # h["Waveform"] = W
        # @test h_in["Symbol"] == SYM
    end
end;


# tests for add_column
t = Table(a = rand(100), b = rand(100)*u"pF")
c = rand(100)*u"mm"
t_new = let df = DataFrame(t); df.c = c; Table(df) end # add the column c to t

LHDataStore(filename, "cw") do h
    h["Table"] = t
    LegendHDF5IO.add_column(h, "Table", (c = c,))
    @test h["Table"] != t_new # probably because it's cached ?
end
    
LHDataStore(filename) do h
    @test h["Table"] == t_new
end
```

Two things to note:
- `LHDataStore` is not able to write `Symbols` (needed to save SSD simulations) or single `RDWaveforms`. This is something that we might want to add later
- When adding a column, one needs to close the file and re-open it in order to get the new table. Seems like there is some caching going on. This should be written down somewhere (maybe in the docstring of `add_column`?)

Maybe add (part of) these tests to the test scripts, such that new functionalities will always be tested before being merged

fhagemann commented:
The command for reading/writing a `Symbol` should be part of `LegendHDF5IO#main` with the newest commit:
https://github.com/legend-exp/LegendHDF5IO.jl/commit/276e45513626eb86a65e80801d59fa1bc9c7a060

Does that need modifications as well or are we fine here?

oschulz commented:
> LHDataStore is not able to write Symbols (needed to save SSD simulations) or single RDWaveforms. This is something that we might want to add later

@apmypb can you add this quickly?

fhagemann commented:
> > LHDataStore is not able to write Symbols (needed to save SSD simulations) or single RDWaveforms. This is something that we might want to add later
> 
> @apmypb can you add this quickly?

Never mind, the methods for writing `Symbols` are already in there (I just tested the tagged version of LegendHDF5IO, and not the newest commit), everything with `Symbols` works fine.

For single `RDWaveforms`, I don't see the urge of already having this in this PR.

apmypb commented:
Should I set the default to maybe, or to some value like 10_000?
I would set it to `nothing`, since we don't necessarily know the structure of the data of our users. 
So in the end only the user knows the chunk_size suitable for him.
Also, the way I changed the `setindex!` functions, you can also specify chunk_size for a `Table` such that all columns inherit the given `chunk_size`. Should I still set `chunk_size` as a property of `LHDataStore`?

oschulz commented:
Hm, let's set the default to 65536 bytes for now.

oschulz commented:
I think be best approach is to determine chunk size base on "first write", i.e. set chunk size to the size of the arrays, columns, etc. passed to the `setindex` that ultimately creates the datasets. It'll be the responsibility of the user to write the data in suitable chunks.

apmypb commented:
I don't think that `LHDataStore` should have this property, since it only manages `LH5Array`'s. Having this property would also imply that we expect every `HDF5.Dataset` inside this `LHDataStore` to be saved in chunks, which might not be the case for simple text info for example. Keeping the current state would give the user freedom to chunk the dataset he really needs.

My question would be if we should give the user the option to choose the size of chunks and the dimension along which one can extend the dataset, or if only the last dimension of an array is getting chunked (current state). Since the chunking feature only really exists for the purpose of being able to append additional data, we might change the current `append!` implementations to something like `cat` with an argument `dims` determining the dimension along which one can extend the dataset.

oschulz commented:
> I don't think that LHDataStore should have this property, since it only manages LH5Array's. Having this property would also imply that we expect every HDF5.Dataset inside this LHDataStore to be saved in chunks, which might not be the case for simple text info for example.

That's no problem, we'd only apply chunking where it makes sense anyway.

> My question would be if we should give the user the option to choose the size of chunks and the dimension along which one can extend the dataset, or if only the last dimension of an array is getting chunked (current state).

Growing only along the last dimension should be enough even long-term. So we should also stick with `append!`.

apmypb commented:
The chunking feature should work now. 
I also changed the `add_column` function to `extend_datastore` and added `reduce_datastore` which add and remove columns from tables or elements from NamedTuples on disk, changing the datatype accordingly.
I don't understand why some tests seem to fail on windows though...

fhagemann commented on test/test_wrappers.jl:
```diff
@@ -67,23 +67,23 @@ using StatsBase
-            LHDataStore(path, "cw") do f
-                f["tmp"] = nt
+            lh5open(path, "cw") do f
+                f["tmp", 50] = nt
```
This is the syntax to write with a chunk size of 50?

apmypb replied:
The integer value in `setindex!` determines the chunk size of the **last** dimension of this array, or if it's a NamedTuple, of each of the arrays inside it.

fhagemann commented on test/test_wrappers.jl:
```diff
@@ -108,4 +108,32 @@ using StatsBase
+            path = joinpath(tmp, "tmp.lh5")
+            lh5open(path, "cw") do lhd
+                lhd["tbl"] = tbl
+                reduce_datastore(lhd, "tbl/y")
```
I think it fails on Windows because of this. I wouldn't be surprised if the syntax on Windows is `tbl\\y`.

fhagemann replied:
Actually, it was the other way around.. :sweat_smile:

fhagemann commented on test/test_wrappers.jl:
```diff
@@ -108,4 +108,32 @@ using StatsBase
+            path = joinpath(tmp, "tmp.lh5")
+            lh5open(path, "cw") do lhd
+                lhd["nt"] = nt
+                reduce_datastore(lhd, "nt/y")
```
Same thing here for Windows (using `\\` instead of `/`).
Maybe this can be avoided using `joinpath(nt, y)`?

apmypb replied:
`joinpath` didn't work...

fhagemann commented on src/types.jl:
```diff
@@ -305,6 +303,7 @@ julia> lhf["new"] = x
 """
 mutable struct LHDataStore <: AbstractDict{String,Any}
     data_store::HDF5.H5DataStore
+    usechunks::Bool
```
Is this ever used...

fhagemann commented on src/types.jl:
```diff
@@ -345,131 +344,129 @@ end
 Base.show(io::IO, lh::LHDataStore) = show(io, MIME"text/plain"(), lh)
 
+Base.setindex!(output::LHDataStore, v, i, chunk_size=nothing) = begin
+    output.usechunks = !isnothing(chunk_size)
```
... other than here (where it is set)?

oschulz commented:
@apmypb is this good to merge from your side?

CC @theHenks

fhagemann commented:
Do we care about the failing windows tests though?

oschulz commented:
The windows test should definitely be fixed, they're passing on main, so there shouldn't be any problem with HDF5.jl itself.

fhagemann commented:
The problem with Windows was that functions like `joinpath` and `splitdir` are defined with a path separator of `\\` (see [here](https://github.com/JuliaLang/julia/blob/3120989f39bb7ef7863c4aab8ab1227cf71eec66/base/path.jl#L21-L43)), whereas the paths in an HDF5 file are still defined as `/`, no matter the operating system.
I have modified the code such that it is explicitly always using `/` in LegendHDF5IO.jl

oschulz commented on src/LegendHDF5IO.jl:
```diff
@@ -15,7 +15,7 @@ using Unitful
 import HDF5
 import TypedTables
-
+using TypedTables: Table, columns
```
```suggestion
using Tables: columns
using TypedTables: columns
```

apmypb commented on src/types.jl:
```diff
@@ -497,3 +494,82 @@ function lh5open(f::Function, filename::AbstractString, access::AbstractString =
+
+extend the Table `dest` at `lhd[i]` with columns from `src`.
+"""
+function extend_datastore(lhd::LHDataStore, i::AbstractString, 
```
`extend_datastore` also supports NamedTuples. I found it very convenient, to not load in the whole table in order to add just a column or another entry in the NamedTuple.

apmypb replied:
having `dest` implicitly checks whether the data structure which we want to alter is actually a named tuple or a table, but I guess, I can also check it inside the function itself.

oschulz replied:
> I found it very convenient, to not load in the whole table in order to add just a column or another entry in the NamedTuple.

When do we add columns, actually?

fhagemann replied:
I do this a lot, when I have raw data and determine measured pulse amplitudes and want to add them to the `Table`. I find this very convenient.

oschulz replied:
That's actually kind of an anti-pattern, we do this via "horizontal" views across files instead (#37).

In principle, scientific data should not be modified/amended in the same file, once written and closed. But we can leave the functionality in if it's currently in active use.

fhagemann replied:
I don't seem to understand what you mean. 

An example LH5 display of a saved Table called "segBEGe" looks like this:
```
🗂️ HDF5.File: (read-only)
└─ 📂 segBEGe
   ├─ 🏷️ datatype
   ├─ 🔢 Phi
   │  ├─ 🏷️ datatype
   │  └─ 🏷️ units
   ├─ 🔢 R
   │  ├─ 🏷️ datatype
   │  └─ 🏷️ units
   ├─ 🔢 Z
   │  ├─ 🏷️ datatype
   │  └─ 🏷️ units
   ├─ 🔢 chid
   │  └─ 🏷️ datatype
   ├─ 🔢 file_idx
   │  └─ 🏷️ datatype
   ├─ 🔢 samples
   │  └─ 🏷️ datatype
   └─ 🔢 z
      ├─ 🏷️ datatype
      └─ 🏷️ units
```

Right now, each COLUMN of a Table is saved as a (separate) HDF5 object in the file,
so why would add a column break the pattern?

oschulz replied:
No, it wouldn't break the pattern. But from a scientific workflow point of view we should avoid amending tables during the course of data taking and analysis. But it's not a problem to offer the functionality.

apmypb commented on src/types.jl:
```diff
@@ -140,7 +138,7 @@ return a `VectorOfVectors` object where `data` is an `LH5Array`
 """
 LH5Array(ds::HDF5.H5DataStore, 
-::Type{<:AbstractVector{<:AbstractVector{<:RealQuantity}}}) = begin
+::Type{<:AbstractVector{<:AbstractVector}}) = begin
```
Vectors for example. The previous version was not able to read in vectors of vectors of vectors. I could add an additional type parameter T which is `Union{RealQuantity, AbstractVector{RealQuantity}}`

oschulz replied:
Oh, silly me, of course!

apmypb commented:
I've changed `reduce_datastore` to `delete_entry!` and `extend_datastore` to `add_entries!`.
Also, chunking by default is set to false. If the `usechunks` variable is set to `true` the chunk size is set to the length of the last axis of the corresponding array. However, `create_entry` also allows for more fine grained control, by setting the `chunk_size` keyword to some positive integer value.

fhagemann commented:
In order to get this going, I'm merging this into a `dev` branch, which can then go to `main` if approved by @oschulz.
@TheHenks: feel free to try it out

oschulz commented:
Thanks @fhagemann - yes, let's use this on `dev` in the field for a few days before merging into main.

theHenks commented:
I will move on with data production with `dev`. I can provide feedback if it works! Thanks @fhagemann
